### ISSUE TITLE
Provide a 'duk_get_**_or' API?

### ISSUE THREAD

**Issue by harold-b** [svaarala/duktape#1263]
What do you think of replacing the `duk_get_**` API calls with one which allows the user to provide a default value which will be returned in the case that the stack index is invalid or it is an unexpected value type. Then, simply change the current `duk_get_**` API to macros which will in turn call new API with the default argument assigned to them in the current API.

For example, the `duk_get_number` implementation could be changed to:

``` C
DUK_EXTERNAL duk_double_t duk_get_number_or(duk_context *ctx, duk_idx_t index, duk_double_t defvalue ) {
	duk_double_union ret;
	duk_tval *tv;

	DUK_ASSERT_CTX_VALID(ctx);

	ret.d = defvalue;  /* Only line changed */
	tv = duk_get_tval(ctx, index);
	if (tv && DUK_TVAL_IS_NUMBER(tv)) {
		ret.d = DUK_TVAL_GET_NUMBER(tv);
	}

	/*
	 *  Number should already be in NaN-normalized form, but let's
	 *  normalize anyway.
	 */

	DUK_DBLUNION_NORMALIZE_NAN_CHECK(&ret);
	return ret.d;
}
```

Then the current `duk_get_number` could be changed to:
``` C
#define duk_get_number(cx, index) duk_get_number_or((cx), (index), DUK_DOUBLE_NAN)
```

I'm suggesting this because to often I have to check the value type at an index first before obtaining it for optional/default arguments in my native bindings. This would simplify that and avoid branching.

Example:
``` C
// Instead of this:
static duk_ret_t play( duk_context* cx )
{
    // JS Signature play( offset:number /*= 0.0*/, loops:number /*=2*/ ):void

    float offset = 0.0f;
    int   loops  = 2;

    if( duk_is_number( cx, 0 ) )
    	offset = (float)duk_get_number( cx, 0 );

    if( duk_is_number( cx, 1 ) )
    	loops = (int)duk_get_int( cx, 1 );

    return 0;
}

// Instead of this:
static duk_ret_t play( duk_context* cx )
{
    // JS Signature play( offset:number /*= 0.0*/, loops:number /*=2*/ ):void

    float offset = (float)duk_get_number_or( cx, 0, 0.0 );
    int   loops  = (int)duk_get_int_or( cx, 1, 2 );

    return 0;
}
```

`duk_get_***_or` is not necessarily the name I'm suggesting — that's another conversation— I'm rather focusing on the purpose :)

Thoughts?

**Comment by svaarala** [comment]
This would definitely be useful, and it's actually been discussed (and even prototyped) a while back. I think the form/naming used then was:

```c
int port = duk_opt_int(ctx, 2, 8080);
```

I recall vaguely that the prototype wasn't merged because there were some type checking / semantics issues to be worked out (for example what to do if an "int" is required by the value is a non-integer number).

This would be a nice candidate for 2.1.0.

**Comment by harold-b** [comment]
Excellent!
Interesting caveat re: semantic issues. What where your thoughts one overcoming the issue you cited? Should there be 2 different calls in such a case? A strict one or a lenient one, or what else is proposed?

**Comment by harold-b** [comment]
Excellent, thanks for filling me in. I'll assist with wherever I can.

**Comment by svaarala** [comment]
Looking through the API calls for get/require, it's mostly the integer related that are a bit unclear.

That's already the case for the integer calls though. For example, `duk_require_int()` actually means "require number, coerce number to integer for return value but *don't* coerce it on the value stack". It's in essence `duk_require_number_return_as_int()`. It's hard to convey that accurately in the naming. Even so, I think the current semantics may be a bit surprising just by looking at the function name.

But if those semantics are not changed, `duk_opt_(u)int()` should simply be consistent with whatever is already in place.

**Comment by harold-b** [comment]
Coercion, as it is with get/require, seems like the sensible way to go. In the odd case that a user requires a stricter lookup, I think it's justified in that case for them write up a check for a fractional, etc. himself.

**Comment by svaarala** [comment]
As a concrete example, calling Ecmascript code:

```js
server.listen('127.0.0.1', 8080.5);  // mysterious fraction ;)
```

C code:

```c
const char *host = duk_opt_string(ctx, 0, "localhost");
int port = (int) duk_opt_int(ctx, 1, 80);
```

With the argument `8080.5` the discussed behavior would result in `port` being `8080`. This seems like a better default than `80` to me, although one could also argue the argument is invalid and should be rejected altogether and the default be used instead. Throwing an error would be surprising behavior for a "get with default" API call because `duk_get_xxx()` also don't throw.

If this seems reasonable I should be able to complete #1264 pretty quickly.

**Comment by harold-b** [comment]
Yes, I completely agree with your last comment. 

I did mentioned earlier if a 2 API call solution should be implemented (strict/lenient), but I thought about in which case someone, who is already specifically calling the `duk_require/get/opt_int` API, would not want it to be coerced if it happens to be a non-integral number. Which seems completely unreasonable, in which case one would explicitly use the `duk_get_number` API instead and do what they will with it themselves.

Coercion is in line with what is already in place, and also familiar to float to integral casts in C, which drop the fraction. So I agree that this is the way to go.

**Comment by svaarala** [comment]
Alright, I'll work up #1264 tonight and it would be great if you could give it a test drive in your application :)

**Comment by harold-b** [comment]
Ok I'll keep an eye out!

I just noticed you released v2.0.0 which seems to have a nice performance boost across the board. So this is great timing for me to upgrade, and switch to the #1264 branch :)

I'm diffing the debugger.rst so I can upgrade the vscode debugger soon as well.

**Comment by fatcerberus** [comment]
Yep, Duktape 2.0 is noticeably more performant than previous versions and has a bunch of nice new goodies too. :)

**Comment by svaarala** [comment]
@harold-b Going through old issues, there was actually a previous proposal to use `duk_get_string_default()` to provide an explicit default. Maybe that would be a good enough name if we can't come up with a better one-word alternative?

**Comment by svaarala** [comment]
This has been implemented in Duktape 2.1 with duk_get_xxx_default() and with a bit different semantics with duk_opt_xxx().

**Comment by harold-b** [comment]
Thanks for the hard work, Sami.
